drivers/flash/flash_util.c

```diff
@@ -80,3 +87,116 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
 	return -ENOSYS;
 #endif
 }
+
+/* note: caller must first check for positivity (>=0) */
+static inline bool off_add_overflow(off_t offset, off_t size, off_t *result)
+{
+	BUILD_ASSERT((sizeof(off_t) == sizeof(uint32_t)) || (sizeof(off_t) == sizeof(uint64_t)));
+
+	if (sizeof(off_t) == sizeof(uint32_t)) {
+		uint32_t end;
+
+		/* account for signedness of off_t due to lack of s32_add_overflow() */
+		if (u32_add_overflow((uint32_t)offset, (uint32_t)size, &end) || (end > INT32_MAX)) {
+			return true;
+		}
+	} else if (sizeof(off_t) == sizeof(uint64_t)) {
+		uint64_t end;
+
+		/* account for signedness of off_t due to lack of s64_add_overflow() */
+		if (u64_add_overflow((uint64_t)offset, (uint64_t)size, &end) || (end > INT64_MAX)) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/* note: caller must first check for overflow */
+static inline bool flash_ranges_overlap(off_t a_start, off_t a_size, off_t b_start, off_t b_size)
```

Comment:
We need to check more than just that the ranges don't overlap, but that the entirety of source and dest are in different erase regions. Otherwise, you're likely to just end up with a bunch of flash full of the erase value.

Reply:
If I'm interpreting this correctly, I think this basically involves rounding the star addresses down by a write block size and rounding end addresses up by the same.

Note: the caller is responsible for erasing prior to copying, if necessary.

Reply:
@d3zd3z - my intuition says that the start start address of the destination should be a multiple of the write block size, otherwise it would be an error, and then rounding up the end address to a write block size (padded with erase values) should be sufficient. I think, as long as the range does not overlap (with a possibly rounded end address), that should work. I could potentially add a test for each of those conditions as well.

Does that make sense?